File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = max(nums)
        minProd = 1
        maxProd = 1
        
        for n in nums:
            if n == 0:
                minProd = 1
                maxProd = 1
            temp = n * maxProd
            maxProd = max(temp, n * minProd, n)
            minProd = min(temp, n * minProd, n)
            
            if res < maxProd:
                res = maxProd
                
        return res
        

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
import java.util.Arrays;
class Solution {
    public int maxProduct(int[] nums) {
        int res = Arrays.stream(nums).max().getAsInt();
        int curMax = 1;
        int curMin = 1;
        int n = nums.length;
        
        for(int i = 0; i < n; i++) {
            int x = nums[i];
            if (x == 0) {
                curMax = 1;
                curMin = 1;
            }
            
            int a = x * curMax;
            int b = x * curMin;
            curMax = Math.max(Math.max(a, b), x);
            
            curMin = Math.min(Math.min(a, b), x);
                
            if (curMax > res)
                res = curMax;
        }
        
        return res;
    }
}